### ISSUE TITLE
pickle protocol 1 silently fails / causes data loss with `__slots__=()`

### ISSUE THREAD

**Issue by agroszer** [python/cpython#93570]
**Bug report**

pickle protocol 1 silently fails / causes data loss with `__slots__ = ()`

`copyreg._reduce_ex` falls back on falsy `__slots__` to `__dict__` which is then not available.
State becomes None, leading to data loss.

```import pickle

class Slotted():

    __slots__ = 'mapping',

    def __init__(self, mapping):
        self.mapping = mapping

class NotSlotted(Slotted):

    __slots__ = ()


m = NotSlotted({'abc':'cde'})

state = pickle.dumps(m, protocol=2)
loaded = pickle.loads(state)
assert loaded.mapping == m.mapping

state = pickle.dumps(m, protocol=1)
loaded = pickle.loads(state)
assert loaded.mapping == m.mapping
```

fails with

```
  File "fail.py", line 23, in <module>
    assert loaded.mapping == m.mapping
AttributeError: mapping
```

Discovered this with `collections.abc.MappingView` and subclasses

**Your environment**

- CPython versions tested on: 3.9.5
- Operating system and architecture: ubuntu x86_64

**Comment by serhiy-storchaka** [comment]
Already fixed in 3.11. 3.9 is in the security-only fixes mode, so 3.10 is only need a fix.

**Comment by serhiy-storchaka** [comment]
3.10 now only accepts security fixes, and this is not the one.
